In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("../Dataset/real_estate.csv")
df.head()

latitude   longitude  price_per_square_py  py apt_code  dist_from_station
0  37.125541  126.913776          2860.053787  24    GEDAE        1588.406226
1  37.223169  127.026913          1715.073163  59    IHEAD         986.146800
2  37.239029  126.906303          1780.604630  57    FHIBJ        1492.514512
3  37.274433  126.737771          1508.814336  42    JBCIJ        1621.602866
4  37.082670  127.143634          2933.930488  32    EDJCI         945.034519

In [3]:
df.dtypes

latitude               float64
longitude              float64
price_per_square_py    float64
py                       int64
apt_code                object
dist_from_station      float64
dtype: object

In [4]:
df.shape

(777, 6)


Q1. 시공사별 아파트 평수의 평균을 구하려고 한다. 다음 단계에 따라 분석을 수행하고 질문에 답하시오.
단계 1 : 데이터의 첫 번째 행부터 열 번째 행까지의 시공사 코드(apt_code) 리스트를 저장한다. 첫 번째 행의 데이터는 위도, 경도, 평당가, 평수, 시공사 코드, 인근 지하철역과의 거리가 각각 37.125541, 126.913776, 2860.053787, 24, GEDAE, 1588.406226이다.
단계 2 : 시공사 코드(apt_code)가 단계 1에서 구한 리스트에 포함되어 있는 데이터를 전부 제거한다.
단계 3 : 시공사 코드(apt_code)별 평수 평균을 구하고 그 중 가장 큰 값을 구하시오

※ 정답은 반올림하여 소수점 둘째 자리까지 출력하시오. (정답 예시: 2.29)

In [6]:
df.loc[:9, 'apt_code']

0    GEDAE
1    IHEAD
2    FHIBJ
3    JBCIJ
4    EDJCI
5    EHCAD
6    HEBIA
7    HEBIA
8    IIGFI
9    FHIBJ
Name: apt_code, dtype: object

In [7]:
df.head()

latitude   longitude  price_per_square_py  py apt_code  dist_from_station
0  37.125541  126.913776          2860.053787  24    GEDAE        1588.406226
1  37.223169  127.026913          1715.073163  59    IHEAD         986.146800
2  37.239029  126.906303          1780.604630  57    FHIBJ        1492.514512
3  37.274433  126.737771          1508.814336  42    JBCIJ        1621.602866
4  37.082670  127.143634          2933.930488  32    EDJCI         945.034519

In [12]:
apt_codes = df['apt_code'][:10].to_list()
apt_codes

['GEDAE',
 'IHEAD',
 'FHIBJ',
 'JBCIJ',
 'EDJCI',
 'EHCAD',
 'HEBIA',
 'HEBIA',
 'IIGFI',
 'FHIBJ']

In [14]:
df_sub1 = df.copy()
df_sub1.loc[df_sub1['apt_code'].str.contains("GEDAE")==False]

latitude   longitude  price_per_square_py  py apt_code  \
1    37.223169  127.026913          1715.073163  59    IHEAD   
2    37.239029  126.906303          1780.604630  57    FHIBJ   
3    37.274433  126.737771          1508.814336  42    JBCIJ   
4    37.082670  127.143634          2933.930488  32    EDJCI   
5    37.295509  126.825066          1698.052341  32    EHCAD   
..         ...         ...                  ...  ..      ...   
772  37.161852  126.917406          3564.856667  40    IHEAD   
773  37.288945  127.118062          3072.810919  24    GBDGG   
774  37.234408  126.745186          2491.714188  61    EJAIA   
775  37.294494  127.185185          1746.372819  40    EGFBJ   
776  37.004345  127.107611          1667.553347  32    EHCAD   

     dist_from_station  
1           986.146800  
2          1492.514512  
3          1621.602866  
4           945.034519  
5          1565.151207  
..                 ...  
772        1593.663253  
773        1803.257080  
774        1346.821965  
775        1461.672552  
776        1377.100866  

[737 rows x 6 columns]

In [15]:
for code in apt_codes:
    df_sub1 = df_sub1.loc[df_sub1['apt_code'].str.contains(code)==False]
df_sub1.head()

latitude   longitude  price_per_square_py  py apt_code  dist_from_station
10  37.046400  126.837832          3750.582844  40    GBDGG        1493.140743
11  37.215855  127.127967          1945.709193  32    HCCBB         989.515201
12  37.252653  127.107425          2670.280398  32    IDCIC        1802.915110
13  37.132489  126.784801          2246.292722  40    HCCBB         922.157169
14  37.270844  127.064936          2423.397245  24    EGFBJ        1491.398873

In [16]:
df_sub1.shape

(552, 6)

In [18]:
df_sub1.groupby('apt_code')['py'].mean().sort_values()

apt_code
DEACF    30.918919
EJEFI    31.708333
EGFBJ    32.629630
EDEAG    33.000000
AADGJ    33.045455
HCCBB    33.088235
CJCBC    33.761905
GBCBB    33.840000
JACBH    33.923077
CDBFE    34.352941
GBDGG    34.640000
IDCIC    34.666667
DAHBD    34.684211
AEEDH    34.941176
EACGE    35.000000
IFGCB    35.142857
EJAIA    35.148148
DCJJH    35.857143
ICDIG    35.888889
HHGCB    35.967742
HFIAJ    38.000000
DBEGD    38.629630
Name: py, dtype: float64

In [19]:
answer = 38.629630
answer = round(answer, 2)
print(answer)

38.63


Q2. 아파트 매물을 분석하기 위해 군집화를 하려고 한다. 다음 단계에 따라 분석을 수행하고 질문에 답하시오.
단계 1 : 독립 변수들에 대해 K-means 군집 분석을 수행한다. 이 때, 군집 수는 2~9개 중 K-means Silhouette 를 통해 구하고, 이 중 첫번째로 높은 score를 최적의 K로 설정한다.
- 독립 변수(총 5개) : 시공사 코드(apt_code)를 제외한 모든 컬럼

단계 2 : 단계 1에서 최적의 K로 도출한 각 군집의 비율 중, 가장 큰 값을 구하시오

※ 정답은 올림하여 소수점 둘째 자리까지 출력하시오. (정답 예시: 2.2

In [20]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
random_state = 229

In [22]:
df.columns

Index(['latitude', 'longitude', 'price_per_square_py', 'py', 'apt_code',
       'dist_from_station'],
      dtype='object')

In [23]:
col = ['latitude', 'longitude', 'price_per_square_py', 'py',
       'dist_from_station']

In [24]:
scores = []
for k in range(2, 10):
    model = KMeans(n_clusters=k, random_state=random_state).fit(df[col])
    score = silhouette_score(df[col], model.labels_)
    print(k, score)
    scores.append(score)

2 0.634635924729853
3 0.44885775222425267
4 0.4558213655816168
5 0.43755753782339263
6 0.3629999389627252
7 0.36408372786750115
8 0.3786501238713389
9 0.3966255576781226


In [25]:
k = 2
model = KMeans(n_clusters=k, random_state=random_state).fit(df[col])

In [28]:
df_sub2 = df.copy()
df_sub2['label'] = model.labels_
df_sub2.head()

latitude   longitude  price_per_square_py  py apt_code  dist_from_station  \
0  37.125541  126.913776          2860.053787  24    GEDAE        1588.406226   
1  37.223169  127.026913          1715.073163  59    IHEAD         986.146800   
2  37.239029  126.906303          1780.604630  57    FHIBJ        1492.514512   
3  37.274433  126.737771          1508.814336  42    JBCIJ        1621.602866   
4  37.082670  127.143634          2933.930488  32    EDJCI         945.034519   

   label  
0      0  
1      1  
2      1  
3      1  
4      0

In [36]:
df_sub2.groupby('label').size()

label
0    201
1    576
dtype: int64

In [37]:
len(df_sub2)

777

In [38]:
answer = 576/777
answer

0.7413127413127413

In [39]:
answer = 0.75

Q3. 아파트 가격을 예측하는 모델을 만드려고 한다. 다음 단계에 따라 분석을 수행하고 질문에 답하시오.

In [40]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
random_state = 229

In [41]:
df_sub3 = df.copy()

In [42]:
df_sub3['price'] = df_sub3['price_per_square_py']*df_sub3['py']
df_sub3.head()

latitude   longitude  price_per_square_py  py apt_code  dist_from_station  \
0  37.125541  126.913776          2860.053787  24    GEDAE        1588.406226   
1  37.223169  127.026913          1715.073163  59    IHEAD         986.146800   
2  37.239029  126.906303          1780.604630  57    FHIBJ        1492.514512   
3  37.274433  126.737771          1508.814336  42    JBCIJ        1621.602866   
4  37.082670  127.143634          2933.930488  32    EDJCI         945.034519   

           price  
0   68641.290895  
1  101189.316628  
2  101494.463921  
3   63370.202096  
4   93885.775621

In [43]:
train_set, test_set = train_test_split(df_sub3, random_state=random_state, test_size=2/30)

In [44]:
print(train_set.shape)
print(test_set.shape)

(725, 7)
(52, 7)


In [45]:
train_set.head()

latitude   longitude  price_per_square_py  py apt_code  \
301  37.055723  126.885253          1569.575556  40    EHCAD   
325  37.170299  126.758981          1631.638340  32    HHGCB   
178  37.051574  126.755140          3350.798249  24    AEEDH   
443  37.026883  126.921716          1665.869452  24    HFIAJ   
109  37.239580  127.094128          1749.377651  40    DCJJH   

     dist_from_station         price  
301        1464.859132  62783.022231  
325        1371.082457  52212.426875  
178        1669.022820  80419.157982  
443         935.042689  39980.866852  
109        1865.419774  69975.106034

In [46]:
df_sub3.columns

Index(['latitude', 'longitude', 'price_per_square_py', 'py', 'apt_code',
       'dist_from_station', 'price'],
      dtype='object')

In [47]:
col = ['latitude', 'longitude', 'price_per_square_py', 'py',
       'dist_from_station']

In [49]:
model1 = LinearRegression(fit_intercept=True).fit(train_set[col], train_set['price'])

In [50]:
pred1 = model1.predict(test_set[col])

In [51]:
model2 = LinearRegression(fit_intercept=False).fit(train_set[col], train_set['price'])
pred2 = model2.predict(test_set[col])

In [52]:
M1 = ((test_set['price']-pred1)**2).mean()

In [53]:
M2 = ((test_set['price']-pred2)**2).mean()

In [55]:
answer = max(M1, M2)

In [56]:
answer = round(answer, 2)

In [57]:
print(answer)

19007465.67
